code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm

```diff
@@ -0,0 +1,529 @@
+				"unit" = "K",
+			)
+		else
+			singular_tlv += list(
+				"name" = GLOB.meta_gas_info[threshold][META_GAS_NAME],
+				"unit" = "kPa",
+			)
+		singular_tlv += list(
+			"id" = threshold,
+			"warning_min" = tlv.warning_min,
+			"hazard_min" = tlv.hazard_min,
+			"warning_max" = tlv.warning_max,
+			"hazard_max" = tlv.hazard_max,
+		)
+		data["tlvSettings"] += list(singular_tlv)
+
+	if(!locked || user.has_unlimited_silicon_privilege)
+		data["vents"] = list()
+		for(var/obj/machinery/atmospherics/components/unary/vent_pump/vent as anything in my_area.air_vents)
+			data["vents"] += list(list(
+				"refID" = REF(vent),
+				"long_name" = sanitize(vent.name),
+				"power" = vent.on,
+				"checks" = vent.pressure_checks,
+				"excheck" = vent.pressure_checks & ATMOS_EXTERNAL_BOUND,
+				"incheck" = vent.pressure_checks & ATMOS_INTERNAL_BOUND,
+				"direction" = vent.pump_direction,
+				"external" = vent.external_pressure_bound,
+				"internal" = vent.internal_pressure_bound,
+				"extdefault" = (vent.external_pressure_bound == ONE_ATMOSPHERE),
+				"intdefault" = (vent.internal_pressure_bound == 0)
+			))
+		data["scrubbers"] = list()
+		for(var/obj/machinery/atmospherics/components/unary/vent_scrubber/scrubber as anything in my_area.air_scrubbers)
+			var/list/filter_types = list()
+			for (var/path in GLOB.meta_gas_info)
+				var/list/gas = GLOB.meta_gas_info[path]
+				filter_types += list(list("gas_id" = gas[META_GAS_ID], "gas_name" = gas[META_GAS_NAME], "enabled" = (path in scrubber.filter_types)))
+			data["scrubbers"] += list(list(
+				"refID" = REF(scrubber),
+				"long_name" = sanitize(scrubber.name),
+				"power" = scrubber.on,
+				"scrubbing" = scrubber.scrubbing,
+				"widenet" = scrubber.widenet,
+				"filter_types" = filter_types,
+			))
+
+		data["selectedModePath"] = selected_mode.type
+		data["modes"] = list()
+		for(var/mode_path in GLOB.air_alarm_modes)
+			var/datum/air_alarm_mode/mode = GLOB.air_alarm_modes[mode_path]
+			if(!(obj_flags & EMAGGED) && mode.emag)
+				continue
+			data["modes"] += list(list(
+				"name" = mode.name,
+				"desc" = mode.desc,
+				"danger" = mode.danger,
+				"path" = mode.type
+			))
+
+		// forgive me holy father
+		data["panicSiphonPath"] = /datum/air_alarm_mode/panic_siphon
+		data["filteringPath"] = /datum/air_alarm_mode/filtering
+
+	return data
+
+/obj/machinery/airalarm/ui_act(action, params)
+	. = ..()
+
+	if(. || buildstage != AIRALARM_BUILD_COMPLETE)
+		return
+	if((locked && !usr.has_unlimited_silicon_privilege) || (usr.has_unlimited_silicon_privilege && aidisabled))
+		return
+
+	var/mob/user = usr
+	var/area/area = get_area(src)
+	ASSERT(!isnull(area))
+
+	var/ref = params["ref"]
+
+	// Possible machines this can refer to
+	var/obj/machinery/atmospherics/components/unary/vent_pump/vent = isnull(ref) ? null : locate(ref) in area.air_vents
+	var/obj/machinery/atmospherics/components/unary/vent_scrubber/scrubber = isnull(ref) ? null : locate(ref) in area.air_scrubbers
```

Comment:
is locate(null) nonvalid?

Reply:
also you mind just doing an if wrap if so this is too much for one line, specially with the in

Reply:
Its probably an optimization.